src/Microsoft.ML.Recommend/MatrixFactorizationPredictor.cs

```diff
@@ -0,0 +1,403 @@
+            for (int i = 0; i < _k; i++)
+                score += _p[poffset + i] * _q[qoffset + i];
+            return score;
+        }
+
+        public ISchemaBoundMapper Bind(IHostEnvironment env, RoleMappedSchema schema)
+        {
+            Contracts.AssertValue(schema);
+            return new RowMapper(this, schema, new ScoreMapperSchema(OutputType, MetadataUtils.Const.ScoreColumnKind.Regression));
+        }
+
+        private sealed class RowMapper : ISchemaBoundRowMapper
+        {
+            private readonly MatrixFactorizationPredictor _parent;
+            private readonly RoleMappedSchema _inputSchema;
+            private readonly ScoreMapperSchema _outputSchema;
+
+            private readonly int _xColIndex;
+            private readonly int _yColIndex;
+            private readonly string _xColName;
+            private readonly string _yColName;
+
+            public ISchema InputSchema => _inputSchema.Schema;
+
+            public RoleMappedSchema InputRoleMappedSchema => _inputSchema;
+
+            public RowMapper(MatrixFactorizationPredictor parent, RoleMappedSchema schema, ScoreMapperSchema outputSchema)
+            {
+                Contracts.AssertValue(parent);
+                _parent = parent;
+
+                var list = schema.GetColumns(RecommendUtils.XKind);
+                if (Utils.Size(list) != 1)
+                    throw Contracts.Except($"'{RecommendUtils.XKind}' column doesn't exist");
+                _xColName = list[0].Name;
+                _xColIndex = list[0].Index;
+
+                list = schema.GetColumns(RecommendUtils.YKind);
+                if (Utils.Size(list) != 1)
+                    throw Contracts.Except($"'{RecommendUtils.YKind}' column doesn't exist");
+                _yColName = list[0].Name;
+                _yColIndex = list[0].Index;
+
+                CheckInputSchema(schema.Schema, _xColIndex, _yColIndex);
+                _inputSchema = schema;
+                _outputSchema = outputSchema;
+            }
+
+            public ISchema Schema => OutputSchema;
+
+            public ISchema OutputSchema
+            {
+                get { return _outputSchema; }
+            }
+
+            public Func<int, bool> GetDependencies(Func<int, bool> predicate)
+            {
+                for (int i = 0; i < OutputSchema.ColumnCount; i++)
+                {
+                    if (predicate(i))
+                        return col => (col == _xColIndex || col == _yColIndex);
+                }
+                return col => false;
+            }
+
+            public IEnumerable<KeyValuePair<RoleMappedSchema.ColumnRole, string>> GetInputColumnRoles()
+            {
+                yield return RecommendUtils.XKind.Bind(_xColName);
+                yield return RecommendUtils.YKind.Bind(_yColName);
+            }
+
+            private void CheckInputSchema(ISchema schema, int xCol, int yCol)
+            {
+                var type = schema.GetColumnType(xCol);
```

Comment:
>schema [](start = 27, length = 6)

Can this be converted to Contracts.Check or Contracts.Assert? You should be able to get the host from the parent and print out a message in the event of failure.

Reply:
What do you mean? Replacing "throw"s below with Contracts.Check?

---
In reply to: [225356071](https://github.com/dotnet/machinelearning/pull/1263#discussion_r225356071) [](ancestors = 225356071)